content/blog/2017-11-10-ozunconf2017realtime.md

```diff
@@ -0,0 +1,93 @@
+
+![](/img/blog-images/2017-11-10-realtime/p5sound_optimised.gif)
+
+This seemed to be worth a go. I managed to follow enough of [this tutorial](https://www.htmlwidgets.org/develop_intro.html) to have the library called from R. I modified the javascript canvas code to look a little more familiar, and the first iteration of `geom_realtime()` was born.
+
+<iframe width="560" height="315" src="https://www.youtube.com/embed/3mdiCUbgxi0" frameborder="0" gesture="media" allowfullscreen></iframe>
+
+This seemed like enough of an idea that I proposed it on the GitHub Issues for the unconf. It got a bit of attention, which was worrying, because I had no idea what to do with this next. Peter Haitch pointed out that Sean Kross had [already wrapped some of the `p5.js` calls into R calls](https://seankross.com/2017/08/11/Beyond-Axes-Simulating-Systems-with-Interactive-Graphics.html) with his `p5` package, so this seemed like a great place to start. It's quite a clever way of doing it too; it involves re-writing the `javascript` which `htmlwidgets` calls on each time you want to do something.
+
+Fast forward to the unconf and a decent number of people gathered around a little slip of paper with `geom_realtime()` written on it. I had to admit to everyone that the `ggplot2` aspect of my demo was a sham (it's surprisingly easy to draw a canvas in just the right shade of grey with white gridlines), but people stayed, and we got to work seeing what else we could do with the idea. We came up with some suggestions for input sources, some different plot types we might like to support, and set about trying to understand what Sean's package actually did.
+
+As it tends to work out, we had a great mix of people with different experience levels in different aspects of the project; some who knew how to make a package, some who knew how to work with `javascript`, some who knew how to work with `websockets`, some who knew about realtime data sources, and some who knew about nearly none of these things (✋ that would be me). If everyone knew every aspect about how to go about a project I suspect the endeavour would be a bit boring. I love these events because I get to learn so much about so many different topics.
+
+I shared my demo script and we deconstructed the pieces. We dug into the inner workings of the `p5` package and started determining which parts we could siphon off to meet our own needs. One of the aspects that we wanted to figure out was how to simulate realtime data. This could be useful both for testing, and also in the situation where one might want to 're-cast' some time-coded data. We were thankful that Jackson had gone deep-dive into `websockets` and pretty soon (surprisingly soon, perhaps; within the first day) we had examples of (albeit, constructed) real-time (every 100ms) data streaming from a server and being plotted at-speed:
```

Comment:
re "Jackson" here, better to refer to people by full name at first, 'cause we don't know that you'll give full name and link in the end